### PR TITLE
feat: Add a request context to authorization related interfaces.

### PR DESCRIPTION
**Pull request by justinusmenzel**
Issue #, if available: (no issue yet, I can open one)

Description of changes: Adding interface to the authorization code to allow for passing an request context through various authorization functions. The request context contains information from the incoming HTTP request but can also contain additional information collected during the authorization evaluation process.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/authorization.ts [hunk 1]
```diff
@@ -20,44 +20,64 @@ export interface BulkDataAuth {
     importResources?: string[];
 }
 
+/**
+ * Contains information from incoming http request. As well as
+ * additional context information such use case specific auth handlers or
+ * operational information specific to this request.
+ */
+export interface RequestContext {
+    method: string;
```
Comment (rsmayda) [inline src/authorization.ts]: Is this URL VERB? i.e. POST, GET, etc? or is this the operation? 'create', 'update', etc

Comment (justinusmenzel) [reply src/authorization.ts]: I'd say this is the HTTP verb, like POST, GET etc.

File: src/authorization.ts [hunk 2]
```diff
@@ -20,44 +20,64 @@ export interface BulkDataAuth {
...
 export interface WriteRequestAuthorizedRequest {
     userIdentity: KeyValueMap;
+    requestContext?: RequestContext;
     operation: TypeOperation;
     resourceBody: any;
 }
 
 export interface GetSearchFilterBasedOnIdentityRequest {
     userIdentity: KeyValueMap;
+    requestContext?: RequestContext;
```
Comment (justinusmenzel) [inline src/authorization.ts:82]: One thing I noticed is that the code calling `getSearchFilterBasedOnIdentity` for example:
https://github.com/awslabs/fhir-works-on-aws-routing/blob/mainline/src/router/handlers/resourceHandler.ts#L68
doesn't seem to have access to `express.Response.locals` where the `RequestContext` instance would be kept. 
So without a lot of changes in `fhir-works-on-aws-routing` to pass `requestContext` around or exposing `express.Response.locals` this approach isn't really working. 
The easiest way, with the least amount of changes is adding the `requestContext` to the `userIdentity` in either the implementation of [verifyAccessToken](https://github.com/awslabs/fhir-works-on-aws-interface/blob/mainline/src/authorization.ts#L74) or in the [authz handler](https://github.com/awslabs/fhir-works-on-aws-routing/blob/mainline/src/app.ts#L71)

File: src/authorization.ts (hunk 1 above)
Comment (rsmayda) [reply src/authorization.ts]: I'd suggest changing the name to verb and making it stricter here seems to be a list of them all: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods

```suggestion
    verb: 'CONNECT' |  'DELETE' | 'GET'| 'HEAD'| 'OPTIONS'|'PATCH'|'POST'|'PUT'|'TRACE';

```

File: src/authorization.ts (hunk 2 above)
Comment (rsmayda) [reply src/authorization.ts:82]: Good research glad you are bringing it up now! I know it is a bit more painful but I would prefer to make it explicit the requestContext is being passed in. It looks like we did something similar for the userIdentity: https://github.com/awslabs/fhir-works-on-aws-routing/blob/77d3f78e916df6208d603d03198a7a5391c5d8ff/src/router/routes/genericResourceRoute.ts#L141

File: src/authorization.ts (hunk 1 above)
Comment (justinusmenzel) [reply src/authorization.ts]: Great suggestion! I'll change that.

File: src/authorization.ts (hunk 2 above)
Comment (justinusmenzel) [reply src/authorization.ts:82]: OK, good to know which path to pursue. I'll start working getting those changes into `fhir-works-on-aws-routing`.

**Review by rsmayda — DISMISSED** [review]
Looks good on my end will have 1 other Engineer to review as well

File: src/authorization.ts [hunk 3]
```diff
@@ -20,44 +20,66 @@ export interface BulkDataAuth {
     importResources?: string[];
 }
 
+export type VerbType = 'CONNECT' | 'DELETE' | 'GET' | 'HEAD' | 'OPTIONS' | 'PATCH' | 'POST' | 'PUT' | 'TRACE';
+
+/**
+ * Contains information from incoming http request. As well as
+ * additional context information such use case specific auth handlers or
```
Comment (nguyen102) [inline src/authorization.ts]: Nitpick: I think we're missing an `as` in the comments.

"...information such `as` use case specific`

**Review by nguyen102 — DISMISSED** [review]
One small comment, otherwise looks good.

File: src/authorization.ts (hunk 3 above)
Comment (justinusmenzel) [reply src/authorization.ts]: Done.

File: src/authorization.ts [hunk 4]
```diff
@@ -103,4 +125,4 @@ export interface Authorization {
      * @returns Search filter
      */
     getSearchFilterBasedOnIdentity(request: GetSearchFilterBasedOnIdentityRequest): Promise<SearchFilter[]>;
-}
+}
```
Comment (rsmayda) [inline src/authorization.ts]: Minor linting required -- locally you can run `yarn run release` to test against the given linter & unit tests

Comment (justinusmenzel) [reply src/authorization.ts]: Fixed. (my IDE was removing newlines at end of file. Changed the config.)

**Review by rsmayda — APPROVED** [review]

**Pull request merged**